src/Core/Log.re

```diff
@@ -4,77 +4,143 @@
+  let bind = (o, f) =>
+    switch (o) {
+    | Some(x) => f(x)
+    | None => None
+    };
 
-let enablePrinting = () => {
-  canPrint := true;
+  let iter = f =>
+    fun
+    | Some(x) => f(x)
+    | None => ();
 };
 
-let enableDebugLogging = () => {
-  debugLogging := true;
+module Constants = {
+  let namespaceStyle = Fmt.(styled(`Fg(`Hi(`Yellow)), string));
+
+  module Env = {
+    let logFile = "ONI2_LOG_FILE";
+    let debug = "ONI2_DEBUG";
+  };
 };
 
-let isDebugLoggingEnabled = () => debugLogging^;
+type msgf('a, 'b) = (format4('a, Format.formatter, unit, 'b) => 'a) => 'b;
+
+let namespaceTag = Logs.Tag.def("namespace", Format.pp_print_string);
+
+let pp_namespace = ppf =>
+  Option.iter(Fmt.pf(ppf, "[%a]", Constants.namespaceStyle));
+
+let fileReporter =
+  switch (Sys.getenv_opt(Constants.Env.logFile)) {
+  | Some(path) =>
+    let channel = open_out(path);
+    Printf.fprintf(channel, "Starting log file.\n%!");
+    let ppf = Format.formatter_of_out_channel(channel);
+
+    Logs.{
+      report: (_src, level, ~over, k, msgf) => {
+        let k = _ => {
+          over();
+          k();
+        };
+
+        msgf((~header=?, ~tags as _=?, fmt) => {
+          Format.kfprintf(
+            k,
+            ppf,
+            "%a@[" ^^ fmt ^^ "@]@.",
+            pp_header,
+            (level, header),
+          )
+        });
+      },
+    };
+
+  | None => Logs.nop_reporter
+  };
 
-let print = msg => canPrint^ ? print_endline(msg) : ();
+let consoleReporter =
+  Logs.{
+    report: (_src, level, ~over, k, msgf) => {
+      let k = _ => {
+        over();
+        k();
+      };
 
-let prerr = msg => canPrint^ ? prerr_endline(msg) : ();
+      msgf((~header=?, ~tags=?, fmt) => {
+        let namespace = Option.bind(tags, Logs.Tag.find(namespaceTag));
 
-let fileChannel =
-  switch (Sys.getenv_opt("ONI2_LOG_FILE")) {
-  | Some(v) =>
-    let oc = open_out(v);
-    Printf.fprintf(oc, "Starting log file.\n");
-    flush(oc);
-    Some(oc);
-  | None => None
+        Format.kfprintf(
+          k,
+          Format.err_formatter,
+          "%a%a@[" ^^ fmt ^^ "@]@.",
+          Logs_fmt.pp_header,
+          (level, header),
+          pp_namespace,
+          namespace,
+        );
+      });
+    },
   };
 
-let _ =
-  switch (Sys.getenv_opt("ONI2_DEBUG"), Sys.getenv_opt("ONI2_LOG_FILE")) {
-  | (Some(_), _) => debugLogging := true
-  | (_, Some(_)) => debugLogging := true
-  | _ => ()
+let reporter =
+  Logs.{
+    report: (src, level, ~over, k, msgf) => {
+      let kret = consoleReporter.report(src, level, ~over=() => (), k, msgf);
+      fileReporter.report(src, level, ~over, () => kret, msgf);
+    },
   };
 
-let logCore = (~error=false, msg) => {
-  switch (fileChannel) {
-  | Some(oc) =>
-    Printf.fprintf(oc, "%s\n", msg);
-    flush(oc);
-  | None =>
-    switch (error) {
-    | false => print(msg)
-    | true => prerr(msg)
-    }
+// defaults
+let () = {
+  Fmt_tty.setup_std_outputs(~style_renderer=`Ansi_tty, ());
```

Comment:
Does this send anything to `stdout`? We have to be careful on Windows - we can actually get a crash by emitting to OCaml's stdout if the console isn't attached (it seems like a bug... `printf` from C is fine, but eventually OCaml `print_endline` hits a limit where it crashes if we log too much w/o an attached console)

Reply:
I don't think it does. From what I can see it just sets some attributes used later to determine how to print.